test/app.test.js

```diff
@@ -1,10 +1,250 @@
-    expect(Array.isArray(getNearestShops({
-      lat: 0,
-      lng: 0,
-    }))).toBe(true);
+import { getNearestShops } from "../src/app";
+import DeltaCoffeeShop from "../src/deltaCoffeeShop";
+import DeltaCoffeeShopList from "../src/deltaCoffeeShopList";
+import SortedDeltaCoffeShopList from "../src/sortedDeltaCoffeeShopList";
+import errorCodes from "../src/errorcodes.js";
+import { ERROR_TOKEN, errorResponse } from "../src/utils.js";
+
+describe("App", () => {
+  it("should return an array when the input is valid", async () => {
+    const f = await getNearestShops({ x: 0, y: 0 });
+    expect(Array.isArray(f)).toBe(true);
+  });
+
+  it("should return an array even if the input is invalid", async () => {
+    const f = await getNearestShops({ x: 0, y: "abc" });
+    expect(Array.isArray(f)).toBe(true);
+  });
+
+  it("should return an array with the error code if the input is invalid", async () => {
+    const f = await getNearestShops({ x: 0, y: "abc" });
+    expect(f.length == 2).toBe(true);
+    expect(f[0]).toBe(ERROR_TOKEN);
+    expect(f[1]).toBe(errorCodes.INVALID_INPUT_ARGS_NAN);
+  });
+
+  it("should return an array with three coffee shops, containing the name and the delta value", async () => {
+    const f = await getNearestShops({ x: 47.6, y: -122.4 });
+    expect(f.length == 3).toBe(true);
+    f.forEach(function (coffeeShop) {
+      expect(coffeeShop[0] === ERROR_TOKEN).toBe(false);
+      expect(typeof coffeeShop.name).toBe("string");
+      expect(typeof coffeeShop.delta).toBe("number");
+    });
+  });
+});
+
+describe("Utils", () => {
+  it("should return the error response in the expected format", () => {
+    let resp = errorResponse(errorCodes.GENERIC_ERROR);
+    let expectedResp = [ERROR_TOKEN, errorCodes.GENERIC_ERROR];
+    expect(resp[0]).toBe(expectedResp[0]);
+    expect(resp[1]).toBe(expectedResp[1]);
+
+    resp = errorResponse(errorCodes.INVALID_INPUT_ARGS_NAN);
+    expectedResp = [ERROR_TOKEN, errorCodes.INVALID_INPUT_ARGS_NAN];
+    expect(resp[0]).toBe(expectedResp[0]);
+    expect(resp[1]).toBe(expectedResp[1]);
+
+    resp = errorResponse(errorCodes.UNAUTHORIZED);
+    expectedResp = [ERROR_TOKEN, errorCodes.UNAUTHORIZED];
+    expect(resp[0]).toBe(expectedResp[0]);
+    expect(resp[1]).toBe(expectedResp[1]);
+
+    resp = errorResponse(errorCodes.UNACCEPTABLE_ACCEPT_FORMAT);
+    expectedResp = [ERROR_TOKEN, errorCodes.UNACCEPTABLE_ACCEPT_FORMAT];
+    expect(resp[0]).toBe(expectedResp[0]);
+    expect(resp[1]).toBe(expectedResp[1]);
+
+    resp = errorResponse(errorCodes.SERVICE_UNAVAILABLE);
+    expectedResp = [ERROR_TOKEN, errorCodes.SERVICE_UNAVAILABLE];
+    expect(resp[0]).toBe(expectedResp[0]);
+    expect(resp[1]).toBe(expectedResp[1]);
+
+    resp = errorResponse(errorCodes.TIMEOUT);
+    expectedResp = [ERROR_TOKEN, errorCodes.TIMEOUT];
+    expect(resp[0]).toBe(expectedResp[0]);
+    expect(resp[1]).toBe(expectedResp[1]);
+  });
+});
+
+function determineDistance(x1, x2, y1, y2) {
```

Comment:
Here we see that this is a copy of the implementation to the test. Fundamentally it means that this is testing something against itself. DO you see a problem with this approach?

Reply:
Since the code is the same, we are basically not testing that specific code section, since testing against itself will always pass